contexts/AuthzContext.tsx

```diff
@@ -0,0 +1,137 @@
+  id: string
+}
+
+type AuthParams = [AuthSignable, unknown]
+
+export type ProposalKey = {
+  address: string
+  keyId: number
+  sequenceNum: number
+}
+
+type AuthSignable = {
+  addr: string
+  args: string[]
+  cadence: string
+  f_type: "Signable"
+  f_vsn: string
+  keyId: number
+  message: string
+  roles: Record<string, boolean>
+  interaction: unknown
+  voucher: {
+    arguments: string[]
+    authorizers: string[]
+    cadence: string
+    computeLimit: number
+    payer: string
+    payloadSigs: string[]
+    proposalKey: ProposalKey
+    refBlock: string
+  }
+}
+
+type AuthzContextType = {
+  currentUser: Account
+  proposer: Account
+  payer: Account
+  authorizers: Account[]
+  roles: Record<string, boolean>
+  proposalKey: ProposalKey
+  args: string[]
+  cadence: string
+  computeLimit: number
+  refBlock: string
+  message: string
+  id: string
+  appTitle: string
+  appIcon: string
+}
+
+export const AuthzContext = createContext<AuthzContextType>({
+  currentUser: {} as Account,
+  proposer: {} as Account,
+  payer: {} as Account,
+  authorizers: [],
+  roles: {},
+  proposalKey: {
+    address: "",
+    keyId: 0,
+    sequenceNum: 0,
+  },
+  args: [],
+  cadence: "",
+  computeLimit: 0,
+  refBlock: "",
+  message: "",
+  id: "",
+  appTitle: "",
+  appIcon: "",
+})
+
+export function AuthzContextProvider({children}: {children: React.ReactNode}) {
+  const [signable, setSignable] = useState<AuthSignable | null>(null)
+  const [id, setId] = useState<string | null>(null)
+  const {data: accountsData} = useAccounts()
+
+  useEffect(() => {
+    function callback({data}: {data: AuthReadyData}) {
+      if (data == null) return
+      if (data.jsonrpc != "2.0") return
+      if (data.method != "fcl:sign") return
+      const [signable] = data.params
+      delete signable.interaction
+      setId(data.id)
+      setSignable(signable)
+    }
+
+    window.addEventListener("message", callback)
+
+    reply("FCL:FRAME:READY")
+
+    return () => window.removeEventListener("message", callback)
+  }, [])
+
+  const accounts = useMemo(() => {
+    if (!accountsData) return {}
+    const hash: Record<string, Account> = {}
+    accountsData.forEach(account => (hash[account.address] = account))
+    return hash
+  }, [accountsData])
+
+  if (!signable || !id || Object.entries(accounts).length === 0) return null
+
+  const {addr: currentUserAddress, voucher, roles, message} = signable
+
+  const value = {
+    currentUser: accounts[fcl.withPrefix(currentUserAddress)],
+    proposer: accounts[fcl.withPrefix(voucher.proposalKey.address)],
+    payer: accounts[fcl.withPrefix(voucher.payer)],
+    authorizers: voucher.authorizers.map(
+      authorizer => accounts[fcl.withPrefix(authorizer)]
+    ),
+    roles,
+    proposalKey: voucher.proposalKey,
+    args: voucher.arguments,
+    cadence: voucher.cadence,
+    computeLimit: voucher.computeLimit,
+    refBlock: voucher.refBlock,
+    message,
+    id,
+    appTitle: "Test Harness",
```

Comment:
@gregsantos is there a way to query the app for its title and icon, or should we save them to the emulator the authn step?

Reply:
It would be available via `config`, If the app adds these details

`fcl.config().put("app.detail.title", "Test Harness").put("app.detail.icon", "https://placekitten.com/g/200/200")`

Reply:
That works for the authn page, but the authz page doesn't receive a FCL:AUTHN:CONFIG message. If there's no way to query the connected app's values, we could store them in the session or the emulator when FCL:AUTHN:CONFIG is received on authn